### PR TITLE
Adding slots information to slurm

### PR DESCRIPTION
**Pull request by lukeseawalker**
On host addition, the slurm configuration is changed to add the number
of slots for the compute node. This information is coming from the sqs
message sent by the compute node itself.

In order to allow to submit more than one job per node, the scheduler
needs also to be configured to use CPUs as consumable resource. This is
done setting the params SelectType and SelectTypeParameters as described
into https://slurm.schedmd.com/slurm.conf.html

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

Test Result using m4.10xlarge (40 CPUs) for Compute nodes.
Without patch, only one job per node. Submitting 5 jobs, only 2 of them go running:
```
$ for i in {1..5}; do sbatch —wrap="sleep 10; echo $(date) $(hostname)"; done
Submitted batch job 2
Submitted batch job 3
Submitted batch job 4
Submitted batch job 5
Submitted batch job 6
$ squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
                1.  compute     wrap ec2-user PD       0:00      1 (Resources)
                2.  compute     wrap ec2-user PD       0:00      1 (Priority)
                3.  compute     wrap ec2-user PD       0:00      1 (Priority)
                4.  compute     wrap ec2-user  R       0:02      1 ip-172-31-2-91
                5.  compute     wrap ec2-user  R       0:02      1 ip-172-31-10-45
```

With this patch, multiple job per node. Submitting 5 jobs, all of them go running:
```
$ for i in {1..5}; do sbatch —wrap="sleep 10; echo $(date) $(hostname)"; done
Submitted batch job 382
Submitted batch job 383
Submitted batch job 384
Submitted batch job 385
Submitted batch job 386
[ec2-user@ip-172-31-6-97 ~]$ squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)

            1.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            2.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            3.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            4.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            5.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
```

With this patch, multiple job per node. Submitting 10 jobs on a cluster with 5 nodes, and requesting 20 "slots" per node. All of them go running, distributed on all 5 nodes:

```
$ for i in {1..10}; do sbatch —ntasks-per-node=20 —wrap="sleep 10; echo $(date) $(hostname)"; done
Submitted batch job 372
Submitted batch job 373
Submitted batch job 374
Submitted batch job 375
Submitted batch job 376
Submitted batch job 377
Submitted batch job 378
Submitted batch job 379
Submitted batch job 380
Submitted batch job 381
[ec2-user@ip-172-31-6-97 ~]$ squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)

            1.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            2.  compute     wrap ec2-user  R       0:01      1 ip-172-31-3-125
            3.  compute     wrap ec2-user  R       0:01      1 ip-172-31-5-34
            4.  compute     wrap ec2-user  R       0:01      1 ip-172-31-5-34
            5.  compute     wrap ec2-user  R       0:01      1 ip-172-31-8-8
            6.  compute     wrap ec2-user  R       0:01      1 ip-172-31-8-8
            7.  compute     wrap ec2-user  R       0:01      1 ip-172-31-8-52
            8.  compute     wrap ec2-user  R       0:01      1 ip-172-31-8-52
            9.  compute     wrap ec2-user  R       0:01      1 ip-172-31-14-69
            10.  compute     wrap ec2-user  R       0:01      1 ip-172-31-14-69
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sqswatcher/plugins/slurm.py [hunk 1]
```diff
@@ -83,52 +83,55 @@ def __readNodeList():
...
+                    partitions_line = slurm_config.next()
+                    node_names_line_items = node_names_line.split(' ')
+                    if len(node_list[partition_name]) > 0:
+                        new_file.write('NodeName=' + ','.join(node_list[partition_name]) + " Procs=" + str(slots) + " " + ' '.join(node_names_line_items[2:]))
                     else:
-                        new_file.write("#NodeName= Procs=1 State=UNKNOWN\n")
-                    items = partitions.split(' ')
-                    node_line = items[1].split('=')
-                    new_file.write(items[0] + " " + node_line[0] + '=dummy-' + partition + ',' + ','.join(node_list[partition]) + " " + ' '.join(items[2:]))
+                        new_file.write("#NodeName= Procs=" + str(slots) + " State=UNKNOWN\n")
```
Comment (sean-smith) [inline sqswatcher/plugins/slurm.py]: Use string interpolation... no type casting required `"#NodeName= Procs=%s State=UNKNOWN\n" % slots `

File: sqswatcher/plugins/slurm.py [hunk 2]
```diff
@@ -83,52 +83,55 @@ def __readNodeList():
...
+                    partitions_line_items = partitions_line.split(' ')
+                    new_file.write(partitions_line_items[0] + ' Nodes=dummy-' + partition_name + ',' + ','.join(node_list[partition_name]) + " " + ' '.join(partitions_line_items[2:]))
```
Comment (sean-smith) [inline sqswatcher/plugins/slurm.py]: Same here

File: sqswatcher/plugins/slurm.py (hunk 1 above)
Comment (lukeseawalker) [reply sqswatcher/plugins/slurm.py]: done

File: sqswatcher/plugins/slurm.py (hunk 2 above)
Comment (lukeseawalker) [reply sqswatcher/plugins/slurm.py]: done

**Pull request merged**